libp2p/network/swarm.py

```diff
@@ -731,6 +863,298 @@ def remove_conn(self, swarm_conn: SwarmConn) -> None:
+                del self.health_data[peer_id]
+
+    # Health Monitoring Methods
+
+    def _start_health_monitoring(self) -> None:
+        """Start health monitoring for all connections."""
+        if not self.connection_config.enable_health_monitoring:
+            return
+
+        logger.debug("Starting connection health monitoring")
+
+        # Initialize health data for existing connections
+        for peer_id, connections in self.connections.items():
+            if peer_id not in self.health_data:
+                self.health_data[peer_id] = {}
+
+            for conn in connections:
+                if conn not in self.health_data[peer_id]:
+                    self.health_data[peer_id][conn] = create_default_connection_health()
+
+    async def _monitor_connections_health(self) -> None:
+        """Periodically monitor all connection health."""
+        logger.debug("Connection health monitoring started")
+
+        while True:
+            try:
+                await trio.sleep(self.health_config.health_check_interval)
+
+                for peer_id, connections in list(self.connections.items()):
+                    for conn in list(connections):
+                        await self._check_connection_health(peer_id, conn)
+
+            except trio.Cancelled:
+                logger.debug("Connection health monitoring cancelled")
+                break
+            except Exception as e:
+                logger.error(f"Health monitoring error: {e}")
+
+    async def _check_connection_health(self, peer_id: ID, conn: INetConn) -> None:
+        """Check individual connection health."""
+        try:
+            # Measure ping latency
+            start_time = trio.current_time()
+            ping_success = await self._ping_connection(conn)
+            latency = (trio.current_time() - start_time) * 1000  # Convert to ms
+
+            # Update health data
+            if peer_id not in self.health_data:
+                self.health_data[peer_id] = {}
+
+            if conn not in self.health_data[peer_id]:
+                self.health_data[peer_id][conn] = create_default_connection_health()
+
+            health = self.health_data[peer_id][conn]
+            health.update_ping_metrics(latency, ping_success)
+            health.update_stream_metrics(len(conn.get_streams()))
+
+            # Check if connection needs replacement
+            if self._should_replace_connection(peer_id, conn):
+                await self._replace_unhealthy_connection(peer_id, conn)
+
+        except Exception as e:
+            logger.error(f"Error checking connection health: {e}")
+
+    async def _ping_connection(self, conn: INetConn) -> bool:
+        """Ping a connection to check health."""
+        try:
+            # Use a simple stream creation test as ping
+            stream = await conn.new_stream()
+            await stream.close()
+            return True
+        except Exception:
+            return False
```

Comment:
We define a `ping_timeout` in Swarm's `ConnectionConfig`, but it doesn't seem to be used here.

Reply:
@pacrob  Good catch! You're right - the `ping_timeout` was defined in the config but not actually used. I've fixed it by wrapping the ping operation with `trio.fail_after(self.connection_config.ping_timeout)` so it now properly respects the configured timeout value.